### PR TITLE
Upgrade Ruby to v3.1.4

### PR DESCRIPTION
**Pull request by heliocola**
## What type of PR is this? (check all applicable)
- [x] Ruby upgrade

## Description
Upgrade to latest 3.1 series
Ruby 3.0 `EOL: 2024-03-31 (expected)` (ref: https://www.ruby-lang.org/en/downloads/branches/)

## Related Tickets & Documents
None

- Related Issue #
- Closes #

## QA Instructions, Screenshots, Recordings
Not applicable

### UI accessibility checklist
Not applicable

## Added/updated tests?
- [x] No, and this is why: current test suite suffice.

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

**Comment by heliocola** [comment]
@klardotsh : I've created this as draft to avoid extra notifications.
This will need the Ruby 3.1.4 image on `ghcr.io/forem/ruby:3.1.4` (and its sha value).
Once that is ready, this PR can be sent for review.

File: Containerfile.base [hunk 1]
```diff
@@ -1,4 +1,4 @@
-ARG RUBY_VERSION=3.0.6
+ARG RUBY_VERSION=3.1.4
```
Comment (klardotsh) [inline Containerfile.base:1]: This doesn't actually need to be changed, it gets pulled from `.ruby-version` automatically as part of [`scripts/build_base_ruby_image.sh`](/scripts/build_base_ruby_image.sh).

That said, this bump also doesn't hurt anything. Maybe I should add a comment here for clarity for next time...

**Comment by klardotsh** [comment]
Hey again @heliocola! Couple things:

- Automatic base image builds came in as part of #19657, though it [isn't tagging any output right now](https://github.com/forem/forem/actions/runs/5595475639/jobs/10231382998?pr=19776) for this PR. I see *why* I did it that way, but I need to update the workflow to probably do the whole build and tag even on PRs rather than just on integration to the main branch. My assumption had been "oh, we'll update Ruby in one PR and then go through and start using it in a second follow-up PR", but this is clearly an incorrect assumption and I need to refactor the workflow to work in a single swing.
- We're pretty swamped on the infra side of things internally for the next week or two, so I can't give you any particular ETA on merge yet. That said, we'll try to get this through relatively quickly as best as we can: the diff seems small, so if regression testing shows no major warts, it should be a much lighter lift than the last upgrade :)

Thanks for this contribution as always - we'll chime back in when we have a bit more breathing room shortly!

**Comment by klardotsh** [comment]
This is now blocked only by the merge of #19871 which will give me the SHA to change `Containerfile` to base off of, and then we should be able to let CI do its thing and see if everything works when combined.

**Comment by klardotsh** [comment]
ughhhhhhhhhhhhhhhhhhhhhhhhh this github bug, RIGHT.

I merged the PR that this depended on, which deleted the target branch, and target branch merges don't work across forks on github (at least I'm pretty sure), and so instead the PR gets closed with no ability to reopen it. I've had this before.

I'm going to reopen this later tonight with the new `FROM` line I alluded to in the above comment, and we'll get some internal eyes on reviewing it ASAP. see you on the far side, @heliocola

**Comment by klardotsh** [comment]
Although, it's giving me an option to re-open all of a sudden... hang on...

**Comment by klardotsh** [comment]
Blocked by #19875 because I'm silly. I'll reopen this once I have reviews and merge on that hotfix :man_facepalming:

**Comment by heliocola** [comment]
@klardotsh : the PR was in draft to avoid notifying everybody, so maybe that was the issue.
I had to fix some broken tests to get CI green and I still have the branch on my fork.

I will try to get a fresh PR today.

**Pull request closed without merging**